allennlp/modules/elmo_lstm_cell.py

```diff
@@ -0,0 +1,227 @@
+            # pass the index at which the shortest elements of the batch finish,
+            # we stop picking them up for the computation.
+            if self.go_forward:
+                while batch_lengths[current_length_index] <= index:
+                    current_length_index -= 1
+            # If we're going backwards, we are _picking up_ more indices.
+            else:
+                # First conditional: Are we already at the maximum number of elements in the batch?
+                # Second conditional: Does the next shortest sequence beyond the current batch
+                # index require computation use this timestep?
+                while current_length_index < (len(batch_lengths) - 1) and \
+                                batch_lengths[current_length_index + 1] > index:
+                    current_length_index += 1
+
+            # Actually get the slices of the batch which we
+            # need for the computation at this timestep.
+            # shape (batch_size, cell_size)
+            previous_memory = full_batch_previous_memory[0: current_length_index + 1].clone()
+            # Shape (batch_size, hidden_size)
+            previous_state = full_batch_previous_state[0: current_length_index + 1].clone()
+            # Shape (batch_size, input_size)
+            timestep_input = inputs[0: current_length_index + 1, index]
+
+            # Do the projections for all the gates all at once.
+            # Both have shape (batch_size, 4 * cell_size)
+            projected_input = self.input_linearity(timestep_input)
+            projected_state = self.state_linearity(previous_state)
+
+            # Main LSTM equations using relevant chunks of the big linear
+            # projections of the hidden state and inputs.
+            input_gate = torch.sigmoid(projected_input[:, 0 * self.cell_size:1 * self.cell_size] +
+                                       projected_state[:, 0 * self.cell_size:1 * self.cell_size])
+            forget_gate = torch.sigmoid(projected_input[:, 1 * self.cell_size:2 * self.cell_size] +
+                                        projected_state[:, 1 * self.cell_size:2 * self.cell_size])
+            memory_init = torch.tanh(projected_input[:, 2 * self.cell_size:3 * self.cell_size] +
+                                     projected_state[:, 2 * self.cell_size:3 * self.cell_size])
+            output_gate = torch.sigmoid(projected_input[:, 3 * self.cell_size:4 * self.cell_size] +
+                                        projected_state[:, 3 * self.cell_size:4 * self.cell_size])
+            memory = input_gate * memory_init + forget_gate * previous_memory
+
+            # Here is the non-standard part of this LSTM cell; first, we clip the
+            # memory cell, then we project the output of the timestep to a smaller size
+            # and again clip it.
+            memory = torch.clamp(memory, -self.memory_cell_clip_value, self.memory_cell_clip_value)
```

Comment:
You could use the in place `memory.clamp_` version, it will be faster if it avoids a new allocation and copy. Same with the other clamp lower down.